### ISSUE TITLE
feat(loss): _extra_loss hook for MSELoss subclasses (regularisers, aux terms)

### ISSUE THREAD

**Issue by zazabap** [zazabap/pdft#17]
## Status

**Spec finalised 2026-05-12.** The downstream regulariser was refined from *uniform* L2-to-identity to **block-structure-aware** L2-to-identity, but the upstream hook requested here is unchanged — purely additive `_extra_loss` on `MSELoss` subclasses. See "Linked downstream artefacts" for the full spec + plan.

## Motivation

`pdft.loss._scalar_loss` currently dispatches on loss type via a closed `isinstance` chain:

```python
def _scalar_loss(pred, target, loss, tensors, m, n, inverse_code) -> Array:
    if isinstance(loss, L1Norm): ...
    if isinstance(loss, MSELoss): ...
    raise TypeError(f"unsupported loss type: {type(loss).__name__}")
```

This means downstream packages cannot define a new loss type (e.g. `BaseLoss + regulariser`) without either modifying upstream pdft or replacing the training loop. Subclassing `MSELoss` does not help, because the `isinstance(loss, MSELoss)` branch ignores subclass-specific state and computes only the base top-k MSE.

**Concrete need.** `pdft-benchmarks` is adding a `qft_identity_regularization` experiment that wants to train a `QFTBasis` with

$$
\mathcal{L}_\text{total}(\theta) = \mathcal{L}_\text{MSE-topK}(\theta) + \lambda \cdot R_\text{block}(\theta)
$$

where

$$
R_\text{block}(\theta) = \sum_{g \in \mathcal{G}_\text{outer}} W \cdot \| T_g - I_g \|_F^2 + \sum_{g \in \mathcal{G}_\text{inner}} \| T_g - I_g \|_F^2
$$

and $I_g$ is the per-gate-kind identity element (Hadamard → $I_2$; controlled-phase → phase 0). The inner/outer mask is matched to `blocked_8`'s 12-vs-60 structural split at $(m, n) = (8, 8)$, $(\text{inner}_m, \text{inner}_n) = (3, 3)$.

The hypothesis is that this block-structure-aware sparsity-inducing bias steers training toward the `blocked_8` basin (32.26 dB on DIV2K-8q) from the dense `qft_identity` basin (31.66 dB). The full design lives in `pdft-benchmarks` at `docs/superpowers/specs/2026-05-12-qft-identity-regularization-design.md`.

This use case wants the *forward* of MSELoss unchanged but an additional term added on top. Without a hook, the options are: (a) reimplement the entire batched training loop in pdft-benchmarks, (b) monkey-patch `_scalar_loss` at experiment-start time, or (c) modify pdft. Option (c) is the smallest and least surprising change.

## Proposed change

Six-line additive hook in `_scalar_loss` (`src/pdft/loss.py`, currently lines 117–123):

```diff
     if isinstance(loss, MSELoss):
         if inverse_code is None:
             raise ValueError("MSELoss requires inverse_code to be provided")
         truncated = topk_truncate(pred, loss.k)
         conj_tensors = [jnp.conj(t) for t in tensors]
         reconstructed = _apply_circuit(conj_tensors, inverse_code, m, n, truncated)
-        return jnp.sum(jnp.abs(target - reconstructed) ** 2)
+        base = jnp.sum(jnp.abs(target - reconstructed) ** 2)
+        # Optional extra-loss hook for MSELoss subclasses (regularisers, aux terms).
+        # Vanilla MSELoss has no `_extra_loss` method → behaviour unchanged.
+        extra_fn = getattr(loss, "_extra_loss", None)
+        if extra_fn is not None:
+            return base + extra_fn(tensors)
+        return base
```

Subclasses opt in by defining a method:

```python
@dataclass(frozen=True)
class MyRegularisedMSELoss(MSELoss):
    lam: float = 0.0
    # ... other fields ...

    def _extra_loss(self, tensors):
        return self.lam * <some scalar computed from tensors>
```

Vanilla `MSELoss` has no such method → `getattr` returns `None` → existing behaviour is bit-identical.

## Design choices (already resolved; left here for reviewer context)

1. **Name: `_extra_loss`** — neutral wrt use case (regularisation vs aux head vs custom prior). Underscore-prefix signals "implementation hook for subclasses, not for external callers."
2. **Signature: `(self, tensors) -> scalar`** — YAGNI on a wider variant. Easy to extend later (default `pred=None, target=None` kwargs) if needed.
3. **`L1Norm` hook?** — Not in this issue. Symmetry argues for adding the same hook to `L1Norm`, but the current use case only needs `MSELoss`. Defer until a use case appears.
4. **Open dispatch table instead of a hook?** — More idiomatic but heavier (per-loss-type registry + state threading). The hook is the smallest incremental change. Revisit if multiple downstream losses materialise.

## Implementation plan (TDD)

Recommended sub-skill for the implementing agent: `superpowers:test-driven-development`.

### Step 1 — Write the failing tests

Append to `tests/test_loss.py`:

```python
from dataclasses import dataclass
import jax.numpy as jnp
from pdft.loss import MSELoss, loss_function
from pdft.bases.circuit.qft import qft_code


def test_mseloss_no_extra_loss_unchanged():
    """Vanilla MSELoss has no _extra_loss; behaviour unchanged."""
    m, n = 2, 2
    code, tensors = qft_code(m, n)
    inv_code, _ = qft_code(m, n, inverse=True)
    pic = jnp.ones((4, 4), dtype=jnp.complex128) / 4.0
    base_loss = float(loss_function(tensors, m, n, code, pic, MSELoss(k=1),
                                     inverse_code=inv_code))
    assert base_loss == base_loss   # finite, no exception


def test_mseloss_extra_loss_hook_adds_term():
    """MSELoss subclass with _extra_loss adds to the scalar."""
    @dataclass(frozen=True)
    class WithExtra(MSELoss):
        def _extra_loss(self, tensors):
            return jnp.asarray(7.0, dtype=jnp.float64)

    m, n = 2, 2
    code, tensors = qft_code(m, n)
    inv_code, _ = qft_code(m, n, inverse=True)
    pic = jnp.ones((4, 4), dtype=jnp.complex128) / 4.0

    base_loss = float(loss_function(tensors, m, n, code, pic, MSELoss(k=1),
                                     inverse_code=inv_code))
    extra_loss = float(loss_function(tensors, m, n, code, pic, WithExtra(k=1),
                                      inverse_code=inv_code))
    assert abs(extra_loss - base_loss - 7.0) < 1e-10


def test_mseloss_extra_loss_uses_tensors():
    """_extra_loss receives the current tensors list and can depend on it."""
    @dataclass(frozen=True)
    class TensorSum(MSELoss):
        def _extra_loss(self, tensors):
            return sum(jnp.sum(jnp.abs(t) ** 2) for t in tensors).real

    m, n = 2, 2
    code, tensors = qft_code(m, n)
    inv_code, _ = qft_code(m, n, inverse=True)
    pic = jnp.zeros((4, 4), dtype=jnp.complex128)  # zero input -> base loss 0

    loss_val = float(loss_function(tensors, m, n, code, pic, TensorSum(k=1),
                                    inverse_code=inv_code))
    expected_reg = float(sum(jnp.sum(jnp.abs(t) ** 2) for t in tensors).real)
    assert abs(loss_val - expected_reg) < 1e-6
```

### Step 2 — Verify the tests fail

```bash
/opt/conda/envs/pdft/bin/python -m pytest tests/test_loss.py::test_mseloss_extra_loss_hook_adds_term -v
```

Expected: FAIL (`extra_loss` equals `base_loss`; hook not yet wired).

### Step 3 — Add the hook to `_scalar_loss`

In `src/pdft/loss.py`, replace the `if isinstance(loss, MSELoss):` block (currently lines 117–123) with:

```python
    if isinstance(loss, MSELoss):
        if inverse_code is None:
            raise ValueError("MSELoss requires inverse_code to be provided")
        truncated = topk_truncate(pred, loss.k)
        conj_tensors = [jnp.conj(t) for t in tensors]  # type: ignore[arg-type]
        reconstructed = _apply_circuit(conj_tensors, inverse_code, m, n, truncated)  # type: ignore[arg-type]
        base = jnp.sum(jnp.abs(target - reconstructed) ** 2)
        # Optional extra-loss hook for MSELoss subclasses (e.g. regularisers).
        # Subclasses define `_extra_loss(tensors) -> scalar`; vanilla MSELoss
        # has no such method and falls through to a 0 contribution.
        extra_fn = getattr(loss, "_extra_loss", None)
        if extra_fn is not None:
            return base + extra_fn(tensors)
        return base
```

### Step 4 — Run the new tests

```bash
/opt/conda/envs/pdft/bin/python -m pytest tests/test_loss.py -v
```

Expected: all three new tests pass; existing `MSELoss` tests still pass.

### Step 5 — Run the full pdft test suite

```bash
/opt/conda/envs/pdft/bin/python -m pytest tests/ -x -q
```

Expected: green.

### Step 6 — Commit (HEREDOC, with trailer)

```bash
git commit -m "$(cat <<'EOF'
feat(loss): add _extra_loss hook for MSELoss subclasses

Lets MSELoss subclasses add an extra scalar term (e.g. a regulariser)
without modifying _scalar_loss dispatch. Vanilla MSELoss has no
_extra_loss method, so behaviour is unchanged. Subclasses define
_extra_loss(tensors) -> scalar; the value is added to the topk-MSE.

Motivated by pdft-benchmarks' qft_identity_regularization experiment
which needs to add a block-structure-aware per-gate L2-to-identity
pull to the training loss; details in pdft-benchmarks
docs/superpowers/specs/2026-05-12-qft-identity-regularization-design.md.

Closes #17.
EOF
)"
```

## Scope

- Touches `src/pdft/loss.py` (one function, ~6 lines added).
- Adds 3 regression tests to `tests/test_loss.py`.
- No public API breakage; the hook is purely additive.

## Out of scope for this issue

- The downstream `BlockMaskedIdentityRegQFTMSELoss` class — lives in `pdft-benchmarks`.
- `L1Norm` hook — deferred until a use case appears.
- Replacing the isinstance dispatch with a `singledispatch` registry — deferred until multiple consumers exist.

## Acceptance criteria

- [ ] Existing `tests/test_loss.py` tests still pass.
- [ ] 3 new tests covering the hook behaviour (no-op when absent; adds term when present; receives tensors).
- [ ] Full pdft test suite (`pytest tests/`) passes.
- [ ] Downstream `pdft-benchmarks` can import `MSELoss`, subclass it, and have `_extra_loss` be invoked end-to-end in `train_basis_batched`.

## Linked downstream artefacts (for context, not in scope)

- pdft-benchmarks spec (revised 2026-05-12 to block-structure-aware regulariser): `docs/superpowers/specs/2026-05-12-qft-identity-regularization-design.md`
- pdft-benchmarks plan (Task 1 = this issue; Tasks 2–N = downstream `identity_reg.py` module, sweep driver, calibration check, figure renderer, writeup): `docs/superpowers/plans/2026-05-12-qft-identity-regularization.md`

The downstream plan is structured for the `superpowers:subagent-driven-development` skill and is the recommended consumer of the hook delivered by this issue.
